File: app/src/main/java/com/coinninja/coinkeeper/util/currency/USDCurrency.java
package com.coinninja.coinkeeper.util.currency;


import android.os.Parcel;
import android.os.Parcelable;

import java.math.BigDecimal;

public class USDCurrency extends BaseCurrency implements FiatCurrency, Parcelable {
    private static final int WHOLE_NUM_MAX = 10;
    private static final int SUB_NUM_MAX = 2;
    public static final String SYMBOL = "$";
    static long MAX_DOLLAR_AMOUNT = Long.MAX_VALUE;
    private String currencyFormat = "$#,##0.00";
    private static final String incrementalFormat = "$#,##0.##";

    public USDCurrency() {
        super();
    }

    public USDCurrency(long initialValue) {
        super(initialValue);
    }

    public USDCurrency(String initialValue) {
        super(initialValue);
    }

    public USDCurrency(double initialValue) {
        super(initialValue);
    }

    public USDCurrency(BigDecimal initialValue) {
        super(initialValue);
    }

    protected USDCurrency(Parcel in) {
        this(in.readLong());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(toLong());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<USDCurrency> CREATOR = new Creator<USDCurrency>() {
        @Override
        public USDCurrency createFromParcel(Parcel in) {
            return new USDCurrency(in);
        }

        @Override
        public USDCurrency[] newArray(int size) {
            return new USDCurrency[size];
        }
    };

    public static void setMaxLimit(USDCurrency currency) {
        try {
            MAX_DOLLAR_AMOUNT = new BTCCurrency(BTCCurrency.MAX_SATOSHI).toUSD(currency).toLong();
        } catch (FormatNotValidException ex) {
            MAX_DOLLAR_AMOUNT = Long.MAX_VALUE;
        }

        if (MAX_DOLLAR_AMOUNT == 0L) {
            MAX_DOLLAR_AMOUNT = Long.MAX_VALUE;
        }
    }

    @Override
    public String getSymbol() {
        return SYMBOL;
    }

    @Override
    public String getFormat() {
        return "#,###.##";
    }

    @Override
    public String getCurrencyFormat() {
        return currencyFormat;
    }

    @Override
    public void setCurrencyFormat(String format) {
        currencyFormat = format;
    }

    @Override
    public int getMaxNumWholeValues() {
        return WHOLE_NUM_MAX;
    }

    @Override
    public String getIncrementalFormat() {
        return incrementalFormat;
    }

    @Override
    public boolean isValid() {
        return toLong() <= getMaxLongValue();
    }

    @Override
    public long getMaxLongValue() {
        return MAX_DOLLAR_AMOUNT;
    }

    @Override
    public int getMaxNumSubValues() {
        return SUB_NUM_MAX;
    }

}

File: common/src/main/java/app/dropbit/commons/currency/USDCurrency.kt
package app.dropbit.commons.currency

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import java.math.BigDecimal
import java.math.RoundingMode
import java.text.DecimalFormatSymbols

class USDCurrency : BaseCurrency, FiatCurrency, Currency, Parcelable {

    constructor() : super()

    constructor(initialValue: Long) : this() {
        value = scale(BigDecimal(initialValue).movePointLeft(maxNumSubValues))
    }

    constructor(initialValue: Double) : this() {
        value = BigDecimal(initialValue).stripTrailingZeros().setScale(this.maxNumSubValues, RoundingMode.HALF_UP)
    }

    constructor(initialValue: BigDecimal) : this() {
        value = initialValue.setScale(maxNumSubValues, RoundingMode.HALF_UP).stripTrailingZeros()
    }

    init {
        decimalSeparator = DecimalFormatSymbols.getInstance().decimalSeparator.toString()
        value = value.stripTrailingZeros()
    }

    internal constructor(parcel: Parcel) : this(parcel.readLong())

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(toLong())
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as USDCurrency

        if (toLong() != other.toLong()) return false

        return true
    }

    override fun hashCode(): Int {
        return 31 * toLong().toInt()
    }

    override val symbol: String = USDCurrency.symbol
    override val format = "#,###.##"
    override val incrementalFormat: String = "$#,##0.##"
    override var currencyFormat = "$#,##0.00"
    override val maxNumSubValues: Int = 2
    override val maxNumWholeValues: Int = 10
    public override val maxLongValue: Long get() = USDCurrency.maxLongValue


    override fun isValid(): Boolean = toLong() <= maxLongValue

    override fun toBTC(conversionValue: Currency): BTCCurrency {
        if (conversionValue.isZero) {
            return BTCCurrency()
        }

        return BTCCurrency(toBigDecimal().divide(conversionValue.toBigDecimal(),
                BTCCurrency.SUB_NUM_MAX, RoundingMode.HALF_UP))
    }

    override fun toUSD(conversionValue: Currency): USDCurrency {
        return this
    }

    override fun toSats(conversionValue: Currency): SatoshiCurrency {
        if (conversionValue.isZero) {
            return SatoshiCurrency()
        }

        return SatoshiCurrency(toBTC(conversionValue).toLong())
    }


    companion object {
        const val symbol = "$"

        @JvmField
        var maxLongValue = Long.MAX_VALUE

        @JvmField
        val CREATOR: Creator<USDCurrency?> = object : Creator<USDCurrency?> {
            override fun createFromParcel(parcel: Parcel): USDCurrency {
                return USDCurrency(parcel)
            }

            override fun newArray(size: Int): Array<USDCurrency?> {
                return arrayOfNulls<USDCurrency?>(size)
            }
        }

        @JvmStatic
        fun setMaxLimit(currency: USDCurrency) {
            try {
                maxLongValue = BTCCurrency(BTCCurrency.maxLongValue).toUSD(currency).toLong()
            } catch (ex: FormatNotValidException) {
                maxLongValue = Long.MAX_VALUE
            }
            if (maxLongValue == 0L) {
                maxLongValue = Long.MAX_VALUE
            }
        }
    }
}